kotlin/internal/repositories/release_repositories.bzl

```diff
@@ -55,3 +60,46 @@ def kotlin_repositories(compiler_release = KOTLIN_CURRENT_COMPILER_RELEASE):
                  "java_stub_template.txt")],
         sha256 = "a618e746e743f3119a9939e60645a02de40149aae9d63201c3cd05706010f6eb",
     )
+
+def _kotlin_compiler_impl(repository_ctx):
+    """Creates the kotlinc repository."""
+    attr = repository_ctx.attr
+
+    repository_ctx.download_and_extract(
+        attr.urls,
+        sha256 = attr.sha256,
+        stripPrefix = "kotlinc",
+    )
+    repository_ctx.file(
+        "WORKSPACE",
+        content = """workspace(name = "%s")""" % attr.name,
+    )
+    repository_ctx.template(
+        "BUILD.bazel",
+        attr._template,
+        substitutions = {
+            "{{.KotlinRules}}": attr.kotlin_rules,
```

Comment:
This is one of the grossest things I've seen, but I get it. Haven't used repository_ctx.template() much myself. (gross in convention, not in concept - it's normal templating in concept)

Reply:
I'm honestly not sure how else to do it.